File: javascript/raytracer/matlab/Camera.m
classdef Camera < Plotable
    %Camera
    
    properties
        pos
        right
        up
        forward
    end
    
    methods
        function obj = Camera(pos, lookAt)
            obj.pos = pos;
            
            obj.forward = lookAt - pos;
            obj.forward = obj.forward / norm(obj.forward);
            
            obj.right = cross([0, 1, 0], obj.forward);
            obj.right = obj.right / norm(obj.right);
            
            obj.up = cross(obj.forward, obj.right);
            obj.up = obj.up / norm(obj.up);
        end
        
        function p = points(obj, width, height, sx, sy)
            if ~exist('sx', 'var')
                sx = width;
            end
            if ~exist('sy', 'var')
                sy = height;
            end
            
            p = zeros(height, width, 3);
            
            halfWidth = width / 2;
            halfHeight = height / 2;
            for i = 0:height-1
                for j = 0:width-1
                    x = (j - halfWidth) / sx;
                    y = (-i + halfHeight) / sy;
                    p(i + 1, j + 1, :) = ...
                        obj.pos + ...
                        x * obj.right + ...
                        y * obj.up + ...
                        obj.forward;
                end
            end
        end
        
        function plot(obj)
            ax = Axes(obj.pos, obj.right, obj.up, obj.forward);
            ax.plot();
            
            p = obj.points(10, 10);
            x = p(:, :, 1);
            y = p(:, :, 2);
            z = p(:, :, 3);
            scatter3(x(:), y(:), z(:));
        end
        
    end
    
end

File: javascript/raytracer/src/Camera.ts
import { Point, Vector } from "./Vector";

import { Ray } from "./Ray";

export class Camera {
    pos: Point;
    right: Vector;
    up: Vector;
    forward: Vector;

    constructor(pos: Point, lookAt: Vector) {
        this.pos = pos;

        this.forward = Vector.unit(
            Vector.sub(lookAt, pos)
        );

        let j = new Vector(0, 1, 0); // i, `j`, k unit vectors
        this.right = Vector.unit(
            Vector.cross(j, this.forward)
        );

        this.up = Vector.unit(
            Vector.cross(this.forward, this.right)
        );
    }

    // todo: add principal point `(ox, oy)`
    *rays(
        width: number,
        height: number,
        sx: number = 0,
        sy: number = 0,
        f: number = 1
    ) {
        if (sx === 0) {
            sx = width;
        }

        if (sy === 0) {
            sy = height;
        }

        let halfWidth = width / 2;
        let halfHeight = height / 2;
        for (let y = 0; y < height; y++) {
            for (let x = 0; x < width; x++) {
                let x_ = (x - halfWidth) / sx;
                let y_ = (-y + halfHeight) / sy;
                let ray = new Ray(
                    this.pos,
                    Vector.sum(
                        Vector.mul(x_, this.right),
                        Vector.mul(y_, this.up),
                        Vector.mul(f, this.forward)
                    )
                );
                
                yield {x, y, ray};
            }
        }
    }
}
